src/core/ddsi/src/q_entity.c

```diff
@@ -2753,12 +2803,20 @@ static bool topickind_qos_match_p_lock (struct entity_common *rd, const dds_qos_
     return false;
   }
   ddsrt_mutex_t * const locks[] = { &rd->qos_lock, &wr->qos_lock, &rd->qos_lock };
+  bool rd_type_lookup, wr_type_lookup;
   const int shift = (uintptr_t) rd > (uintptr_t) wr;
   for (int i = 0; i < 2; i++)
     ddsrt_mutex_lock (locks[i + shift]);
-  bool ret = qos_match_p (rdqos, wrqos, reason);
+  bool ret = qos_match_p (gv, rdqos, rd_typeid, wrqos, wr_typeid, reason, &rd_type_lookup, &wr_type_lookup);
   for (int i = 0; i < 2; i++)
     ddsrt_mutex_unlock (locks[i + shift]);
+
+  if (!ret && (rd_type_lookup || wr_type_lookup))
+  {
+    (void) ddsi_tl_request_type (gv, rd_type_lookup ? rd_typeid : wr_typeid);
```

Comment:
I get why you can do `rd_type_lookup ? rd_typeid : wr_typeid`, but it would be worth a comment.